packages/cli-repl/src/csfle-library-paths.ts

```diff
@@ -0,0 +1,92 @@
+    process.platform === 'win32' ? 'dll' :
+      process.platform === 'darwin' ? 'dylib' : 'so';
+
+export interface CSFLELibraryPathResult {
+  csflePath?: string;
+  expectedVersion?: { version: bigint; versionStr: string };
+}
+
+/**
+ * Figure out the possible shared library paths for the CSFLE shared library
+ * that we are supposed to use.
+ */
+export async function getCSFLELibraryPaths(
+  bus: MongoshBus,
+  pretendProcessExecPathForTesting: string | undefined = undefined): Promise<CSFLELibraryPathResult> {
+  const execPath = pretendProcessExecPathForTesting ?? process.execPath;
+
+  let getCSFLESharedLibraryVersion: typeof import('mongodb-csfle-library-version');
+  try {
+    getCSFLESharedLibraryVersion = require('mongodb-csfle-library-version');
+  } catch (err) {
+    getCSFLESharedLibraryVersion = () => ({ version: BigInt(0), versionStr: '<unknown>' });
+  }
+
+  if (execPath === process.argv[1] || pretendProcessExecPathForTesting) {
+    const bindir = path.dirname(execPath);
+    const execPathStat = await fs.stat(execPath);
+    for await (const libraryCandidate of [
+      // Location of the shared library in the deb and rpm packages
+      path.resolve(bindir, '..', 'lib', `mongosh_csfle_v1.${SHARED_LIBRARY_SUFFIX}`),
+      // Location of the shared library in the zip and tgz packages
+      path.resolve(bindir, `mongosh_csfle_v1.${SHARED_LIBRARY_SUFFIX}`)
+    ]) {
+      try {
+        const permissionsMismatch = await ensureMatchingPermissions(libraryCandidate, execPathStat);
+        if (permissionsMismatch) {
+          bus.emit('mongosh:csfle-load-skip', {
+            csflePath: libraryCandidate,
+            reason: 'permissions mismatch',
+            details: permissionsMismatch
+          });
+          continue;
+        }
+
+        const version = getCSFLESharedLibraryVersion(libraryCandidate);
+        const result = {
+          csflePath: libraryCandidate,
+          expectedVersion: version
+        };
+        bus.emit('mongosh:csfle-load-found', result);
+        return result;
+      } catch (err: any) {
+        bus.emit('mongosh:csfle-load-skip', {
+          csflePath: libraryCandidate,
+          reason: err.message
+        });
+      }
+    }
+  } else {
+    bus.emit('mongosh:csfle-load-skip', {
+      csflePath: '',
+      reason: 'Skipping CSFLE library searching because this is not a single-executable mongosh'
+    });
+  }
+  return {};
+}
+
+async function ensureMatchingPermissions(filename: string, execPathStat: { uid: number, gid: number }): Promise<null | object> {
+  await fs.access(filename, fsConstants.R_OK);
+  const stat = await fs.stat(filename);
+  // On UNIX systems, only load shared libraries if they are coming
+  // from a user we can consider trusted (current user or the one who owns
+  // the mongosh binary to begin with) and they are not writable by other
+  // users. On Windows, no such check is feasible.
+  if (process.platform !== 'win32' &&
```

Comment:
NIT: just a silly thing, I got a bit puzzled by the fact that the function  "returns" `null`, in a good case and an object otherwise. Basically, I understood that we were reporting permission issues Infos only when I got to the usage.

Probably a comment that explains that we return details useful for debugging in case of a mismatch would have helped me.

  
 maybe you could early return here for `win32`

Reply:
Yup, all of that makes sense, done! :+1: